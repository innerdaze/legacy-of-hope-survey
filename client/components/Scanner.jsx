import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Form from 'grommet/components/Form'
import FormField from 'grommet/components/FormField'
import TextInput from 'grommet/components/TextInput'
import Box from 'grommet/components/Box'
import Button from 'grommet/components/Button'

class Scanner extends Component {

  constructor(props) {
    super(props)

    this.state = {
      barcode: null
    }

    this.handleEnterBarcodeManually = this.handleEnterBarcodeManually.bind(this)
    this.handleStartScanClick = this.handleStartScanClick.bind(this)
    this.handleEnterBarcodeFieldChange = this.handleEnterBarcodeFieldChange.bind(this)
  }

  handleEnterBarcodeFieldChange(e) {
    this.setState({
      barcode: e.target.value
    })
  }

  handleEnterBarcodeManually(e) {
    e.preventDefault()
    this.props.onSubmitBarcode(this.state.barcode)
  }

  handleStartScanClick() {
    this.props.scan()
  }

  render() {
    return (
      <Box direction='row'>
        <Form direction='row'>
          <FormField>
            <TextInput
              placeHolder='Enter Barcode'
              onDOMChange={this.handleEnterBarcodeFieldChange}
              />
          </FormField>
          <Button
            primary
            type='submit'
            label='Submit'
            onClick={this.handleEnterBarcodeManually}
            />
        </Form>
      </Box>
    )
  }
}

Scanner.propTypes = {
  onSubmitBarcode: PropTypes.func.isRequired,
  scan: PropTypes.func.isRequired
}

export default Scanner
