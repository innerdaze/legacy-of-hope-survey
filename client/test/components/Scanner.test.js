import React from 'React'
import Scanner from '../../components/Scanner'
let wrapper, onSubmitBarcode, scan

describe('Testing Scanner...', () => {
  beforeEach(() => {
    onSubmitBarcode = jest.fn()
    scan = jest.fn()
    wrapper = shallow(<Scanner onSubmitBarcode={onSubmitBarcode} scan={scan} />)
  })

  describe('Scanner layout', () => {
    test('Check there is a Form with a FormField and a Button', () => {
      expect(wrapper.find('Form')).toHaveLength(1)
      expect(wrapper.find('FormField')).toHaveLength(1)
      expect(wrapper.find('Button')).toHaveLength(1)

      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('Testing on prop', () => {
    test('Expect onSubmitBarcode is called when Button is click', () => {
      wrapper.find('Button').simulate('click', {
        preventDefault() {}
      })
      expect(onSubmitBarcode).toHaveBeenCalled()
    })
  })

  describe('Testing on state', () => {
    test('Expect barcode to change when TextInput change', () => {
      wrapper.find('TextInput').prop('onDOMChange')({
        target: {
          value: 'test'
        }
      })
      expect(wrapper.state().barcode).toEqual('test')
    })
  })
})
