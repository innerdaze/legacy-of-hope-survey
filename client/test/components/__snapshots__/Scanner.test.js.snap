// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Scanner... Scanner layout Check there is a Form with a FormField and a Button 1`] = `
<Box
  announce={false}
  direction="row"
  focusable={true}
  pad="none"
  responsive={true}
  tag="div"
>
  <Form
    compact={false}
    direction="row"
    fill={false}
    pad="none"
  >
    <FormField
      size="medium"
      strong={false}
    >
      <TextInput
        onDOMChange={[Function]}
        placeHolder="Enter Barcode"
      />
    </FormField>
    <Button
      label="Submit"
      method="push"
      onClick={[Function]}
      primary={true}
      type="submit"
    />
  </Form>
</Box>
`;
